### PR TITLE
Issue #14814: refactor extractFullIdent into iterative method

### PR DESCRIPTION
**Pull request by mohitsatr**
part of #14814 

- converted extractFullIdent into iterative method
- added test case for 100% branch coverage 

diffs: https://github.com/checkstyle/checkstyle/pull/15904#issuecomment-2469701116

### CODE REVIEW

**Comment by romani** [comment]
Idea inspections are not happy - https://app.circleci.com/pipelines/github/checkstyle/checkstyle/28706/workflows/3a0e0ffe-4e14-43a7-b30d-933436b18e88/jobs/723126/artifacts

File: src/test/java/com/puppycrawl/tools/checkstyle/api/FullIdentTest.java [hunk 1]
```diff
@@ -210,6 +210,42 @@ private static FullIdent prepareFullIdentWithCoordinates(int columnNo, int lineN
         return FullIdent.createFullIdent(ast);
     }
 
+    @Test
+    public void testNewTypeTag() throws Exception {
+        final DetailAstImpl parent = new DetailAstImpl();
+        parent.setType(TokenTypes.TYPE);
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/FullIdentTest.java:216]: we banned pure junit test methods as they does not contribute to quality as much as we need.


can you make coverage as in testCreateFullIdentBelow2  of this class?

**Review by romani — CHANGES_REQUESTED** [review]
please fix CI

item:

File: src/test/java/com/puppycrawl/tools/checkstyle/api/FullIdentTest.java (hunk 1 above)
Comment (mohitsatr) [reply src/test/java/com/puppycrawl/tools/checkstyle/api/FullIdentTest.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/api/FullIdent.java [hunk 2]
```diff
@@ -70,44 +72,86 @@ public static FullIdent createFullIdent(DetailAST ast) {
...
+        while (!identStack.isEmpty()) {
+            final DetailAST currentAst = identStack.pop();
+
+            final DetailAST nextSibling = currentAst.getNextSibling();
 
             // Here we want type declaration, but not initialization
             final boolean isArrayTypeDeclarationStart = nextSibling != null
                     && (nextSibling.getType() == TokenTypes.ARRAY_DECLARATOR
-                        || nextSibling.getType() == TokenTypes.ANNOTATIONS)
+                    || nextSibling.getType() == TokenTypes.ANNOTATIONS)
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/FullIdent.java:93]: please revert formatting, previous form was better

File: src/main/java/com/puppycrawl/tools/checkstyle/api/FullIdent.java [hunk 3]
```diff
@@ -70,44 +72,86 @@ public static FullIdent createFullIdent(DetailAST ast) {
...
+     * @param full the FullIdent to add to
+     * @param ast the node
+     * @param dotCounter no of dots
+     * @param bracketsExist yes if true
+     * @param isArrayTypeDeclarationStart true if array type declaration start
+     * @return updated value of dotCounter
+     */
+    private static int appendToFull(FullIdent full, DetailAST ast,
+                int dotCounter, boolean bracketsExist, boolean isArrayTypeDeclarationStart) {
+        int lDotCounter = dotCounter;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/FullIdent.java:127]: `lDotCounter` ==> `result`

`l` prefix is weird.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/FullIdent.java [hunk 4]
```diff
@@ -136,8 +180,7 @@ private static boolean isArrayTypeDeclaration(DetailAST arrayDeclarator) {
      * @param ast the type ast we are building a {@code FullIdent} for
      */
     private static void appendBrackets(FullIdent full, DetailAST ast) {
-        final int bracketCount =
-                ast.getParent().getChildCount(TokenTypes.ARRAY_DECLARATOR);
+        final int bracketCount = ast.getParent().getChildCount(TokenTypes.ARRAY_DECLARATOR);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/FullIdent.java:183]: revert or move this not functional formatting to separate commit "minor: formatting"

File: src/test/resources/com/puppycrawl/tools/checkstyle/api/fullident/InputFullIdentLiteralNewCondition.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        int j = 20; // violation, 'Unused local variable'
+        nested obj = new nested() {
+            void method() {
+                j += 50;
+            }
+        };
+        obj.getClass();
+    }
+    class nested {
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/api/fullident/InputFullIdentLiteralNewCondition.java:19]: `Nested` lets keep test code as good formatted if it is not point test.

**Review by romani — CHANGES_REQUESTED** [review]
goods CI is green.
Now we need to list all affected Checks and run regression diff report generation to be sure that have no changes.
https://github.com/checkstyle/test-configs/?tab=readme-ov-file#using-the-configurations
please trigger report generation for all(several if too many) Checks that are potentially affected, use 'all-examples-in-one' bundle.
If you are interested how it works - https://github.com/checkstyle/checkstyle/tree/master/.github/workflows#regression-report-workflow---regression-reportyml


minor changes:

**Comment by romani** [comment]
GitHub, generate report by config from src/test/resources/com/puppycrawl/tools/checkstyle/api/fullident/InputFullIdentLiteralNewCondition.java

**Comment by mohitsatr** [comment]
@romani Like you said, pure units test methods do not contribute to code quality. so at some point in future , we will update these existing tests with input based tests right ? 
https://github.com/checkstyle/checkstyle/blob/aa87e7f9721129678d717063b0340ffb58fb15af/src/test/java/com/puppycrawl/tools/checkstyle/api/FullIdentTest.java#L43-L47

File: src/main/java/com/puppycrawl/tools/checkstyle/api/FullIdent.java (hunk 2 above)
Comment (mohitsatr) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/FullIdent.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/api/FullIdent.java (hunk 4 above)
Comment (mohitsatr) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/FullIdent.java]: done

File: src/test/resources/com/puppycrawl/tools/checkstyle/api/fullident/InputFullIdentLiteralNewCondition.java (hunk 5 above)
Comment (mohitsatr) [reply src/test/resources/com/puppycrawl/tools/checkstyle/api/fullident/InputFullIdentLiteralNewCondition.java]: done

**Comment by mohitsatr** [comment]
GitHub, generate report FullIdent/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for FullIdent/all-examples-in-one

**Comment by romani** [comment]
Expected path is path is this repository https://github.com/checkstyle/test-configs

Just find in that repository check that you need and open folder `all-....` and in readme there will be example of phrase for comment to trigger diff report.


Please list affect Checks to let me help you more precisely.

**Comment by romani** [comment]
>so at some point in future , we will update these existing tests with input based tests right ?

Yes. This is unfortunate branch of tests that were created a long time ago and nobody has time to refactor it.

**Comment by mohitsatr** [comment]
affected Checks: 

1. PackageNameCheck
2. AbstractClassCouplingCheck
3. JavadocMethodCheck
4. UnusedImportCheck
5. RedundantImportCheck
6. ImportOrderCheck
7. ImportControlCheck
8. IllegalImportCheck
9. CustomImportOrderCheck
10. AvoidStaticImportCheck
11. VisibilityModifierCheck
12. VariableDeclarationUsageDistanceCheck
13. PackageDeclarationCheck
14. IllegalTypeCheck
15. IllegalThrowsCheck
16. IllegalInstantiationCheck
17. IllegalCatchCheck
18. EqualsHashCodeCheck
19. CovariantEqualsCheck
20. UncommentedMainCheck
21. UnstableAnnotationCheck
22. SinceFormatCheck
23. ScriptServiceCheck
24. FinalClassCheck
25. UnusedLocalVariableCheck
26. ParameterNumberCheck
27. MissingJavadocTypeCheck
28. MissingJavadocMethodCheck
29. HideUtilityClassConstructorCheck
30. PackageAnnotationCheck
31. MissingDeprecated Check
32. UnstableAnnotationCheck

**Comment by mohitsatr** [comment]
Github, generate report for JavadocMethod/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for UnusedImports/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for RedundantImport/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for ImportOrder/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for IllegalImport/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for AvoidStaticImport/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for VisibilityModifier/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for VariableDeclarationUsageDistance/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for PackageDeclaration/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for IllegalType/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for IllegalThrows/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for IllegalInstantiation/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for IllegalCatch/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for EqualsHashCode/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for CovariantEquals/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for UncommentedMain/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for ParameterNumber/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for MissingJavadocType/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for MissingJavadocMethod/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for PackageAnnotation/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for MissingDeprecated/all-examples-in-one

**Comment by mohitsatr** [comment]
Github, generate report for PackageName/all-examples-in-one

**Review by romani — APPROVED** [review]
as usual in such refactoring, code become less compact but it is ok, lets avoid recursion.

thanks a lot.

next time please run no more that 5 diff reports, to not over stress our CI.
I checked all diff reports, all of them have no diff.
